### PR TITLE
custom audio player component

### PR DESCRIPTION
**Pull request by kiahjh**
closes https://github.com/friends-library/dev/issues/91

### CODE REVIEW

File: apps/evans/components/core/EmbeddedAudio.tsx [hunk 1]
```diff
@@ -1,60 +1,176 @@
...
-    .map((key) => `${encodeURIComponent(key)}=${encodeURIComponent(params[key] ?? ``)}`)
-    .join(`&`);
-
-  return `https://w.soundcloud.com/player/?${query}`;
-}
+  const currentTitle = tracks[currentTrack.current]?.title ?? ``;
+
+  const whilePlaying = useCallback(() => {
+    if (!audio) return;
+    if (audio.currentTime === audio.duration) {
```
Comment (jaredh159) [inline apps/evans/components/core/EmbeddedAudio.tsx:27]: i might be misunderstanding this, but it seems like we're only running this function to progress to the next track, or stop if we're at the end of the last track.  is that right? but we're running the function on every animation frame? i'm not super experienced with requestanimationframe, so maybe this is normal practice, but it seems sort of wasteful to query 60 times per seconds if we're at the end...  is there another way to do this? or maybe it doens't matter. just thinking aloud

File: apps/evans/components/core/EmbeddedAudio.tsx [hunk 2]
```diff
@@ -1,60 +1,176 @@
...
+
+  useEffect(() => {
+    if (isPlaying) {
+      audioRef.current?.play();
+      animationRef.current = requestAnimationFrame(whilePlaying);
+    } else {
+      audioRef.current?.pause();
+      cancelAnimationFrame(animationRef.current ?? 0);
+    }
+  }, [isPlaying, whilePlaying, currentTrack.current]); // eslint-disable-line react-hooks/exhaustive-deps
```
Comment (jaredh159) [inline apps/evans/components/core/EmbeddedAudio.tsx:54]: what dependency is omitted here, and why?

File: apps/evans/components/core/EmbeddedAudio.tsx [hunk 3]
```diff
@@ -1,60 +1,176 @@
...
-    frameBorder="no"
-    src={src(trackId, playlistId, showArtwork)}
-  />
-);
+const EmbeddedAudio: React.FC<Props> = ({ tracks }) => {
+  const [isPlaying, setIsPlaying] = useState(false);
+  const [currentTime, setCurrentTime] = useState(0);
 
-export default EmbeddedAudio;
+  const currentTrack = useRef(0);
```
Comment (jaredh159) [inline apps/evans/components/core/EmbeddedAudio.tsx:18]: why is current track a ref, and not state?

also, slight naming nitpick, could you rename it `currentTrackIndex`?

File: apps/evans/components/core/EmbeddedAudio.tsx [hunk 4]
```diff
@@ -1,60 +1,176 @@
...
+                }}
+              />
+            </div>
+            <span className="text-flgray-500 w-14 flex justify-end">
+              {formatDuration(audio.duration)}
+            </span>
+          </div>
+        </div>
+      ) : (
+        <div>loading</div>
```
Comment (jaredh159) [inline apps/evans/components/core/EmbeddedAudio.tsx:133]: when would we be in a loading state?

File: apps/evans/components/core/EmbeddedAudio.tsx [hunk 5]
```diff
@@ -1,60 +1,176 @@
...
+      )}
+      {tracks.length > 1 &&
```
Comment (jaredh159) [inline apps/evans/components/core/EmbeddedAudio.tsx:135]: can you make a storybook story for the case where there is only one track? that's the most common case, and i can't currently see what it looks like, since we're not pulling real data.

File: apps/evans/components/pages/document/ListenBlock.tsx [hunk 6]
```diff
@@ -7,19 +7,22 @@ import DownloadAudiobook from './DownloadAudiobook';
 import { LANG } from '@/lib/env';
 import EmbeddedAudio from '@/components/core/EmbeddedAudio';
 
-export interface Props {
+export type AudioPart = {
   title: string;
+  url: string;
```
Comment (jaredh159) [inline apps/evans/components/pages/document/ListenBlock.tsx:12]: would it be helpful for the api to tell us the duration in seconds as well? would that prevent the `NaN:NaN` initial load thing i'm seeing now?

File: apps/evans/components/pages/document/ListenBlock.tsx [hunk 7]
```diff
@@ -38,46 +41,31 @@ const ListenBlock: React.FC<Props> = ({ title, numAudioParts, embedId, ...props
...
     >
       <DownloadAudiobook
-        className="mb-8 sm:mb-16 lg:border lg:border-l-0 lg:-mt-12 lg:pt-12 lg:px-12 border-flgray-200 xl:mr-6"
+        className="mb-8 sm:mb-16 lg:border lg:border-l-0 lg:-mt-12 lg:pt-12 lg:px-12 border-flgray-200 xl:mr-6 shrink-0"
         {...props}
         quality={quality}
         setQuality={setQuality}
       />
-      <div className="flex-grow lg:ml-8 xl:max-w-screen-md xl:mx-auto">
+      <div className="flex-grow lg:max-w-[480px] min-[1140px]:max-w-xl min-[1340px]:max-w-3xl lg:mx-auto flex flex-col">
```
Comment (jaredh159) [inline apps/evans/components/pages/document/ListenBlock.tsx:49]: i noticed at 1141 px wide, things don't look so good currently. both the audio and the headline are smashed up against the left side download block.

File: apps/evans/lib/dates.ts [hunk 8]
```diff
@@ -40,3 +40,16 @@ export function spanishShortMonth(short: string): string {
   }
   return short;
 }
+
+export function formatDuration(seconds: number): string {
```
Comment (jaredh159) [inline apps/evans/lib/dates.ts:44]: `lib/dates.ts` seems like a weird place for this file...

also, does it handle edge cases well? can you make a few tests showing that < 60 seconds formats well, and over one hour as well? i've been burned by those sorts of things when i've written formatters like this in the past.

File: apps/evans/pages/getting-started.tsx [hunk 9]
```diff
@@ -106,13 +106,12 @@ const GettingStarted: React.FC<Props> = ({ books, numBooks }) => (
...
-            trackId={LANG === `es` ? 783137959 : 242345955}
-            title={
-              LANG === `es`
-                ? `¿Quienes Eran Los Primeros Cuáqueros?`
-                : `Introduction to the Early Quakers`
-            }
-            showArtwork={false}
+            tracks={[
+              {
+                title: `TODO`,
```
Comment (jaredh159) [inline apps/evans/pages/getting-started.tsx:111]: i'll get you urls for these, so you can leave them as TODO, but you should be able to use the titles that were here before, based on the language.

File: apps/storybook/stories/ListenBlock.stories.tsx [hunk 10]
```diff
@@ -13,9 +13,21 @@ const meta = {
 type Story = StoryObj<typeof meta>;
 
 export const Default: Story = props({
-  title: `Cool book`,
+  tracks: [
```
Comment (jaredh159) [inline apps/storybook/stories/ListenBlock.stories.tsx:16]: let's add a story for one that only has one track as well.

**Review by jaredh159 — CHANGES_REQUESTED** [review]
awesome job, really pleased with this overall, and looking forward to getting it finished and merged, and being done with soundcloud. thanks!

left a bunch of comments, we may need to do a huddle on all of the state/requestAnimationFrame stuff, because i might just be clueless in what is necessary to make all this work.

couple things i didn't mention as comments on _lines_:

1. i feel like the look and feel of the player isn't quite right, it doesn't feel like it fits with the rest of the website. might just be mostly rounded-ness -- the site as it stands now has very rounded buttons, but most other rectangles don't have semi-rounded corners, and so it makes this feel widget feel kindof odd, like a transplant from another website. i'm open to revamping the whole feel of the website at some point, but i'd want to do it wholistically, not piecemeal. can you take a little pass over this, looking at corners and typography and little stuff like that, with an eye to making it feel like it was designed to go with the rest of the site?
2. when i load it up in the vercel preview i see `Nan:Nan`
3. we need to think about the fact that we currently support hi-quality and lo-quality audio. the player should pull the right mp3s based on which one is selected by the user.  this might mean we just pass different mp3 urls in from the parent (changing  nothing in your component), or it might mean your component itself needs to be quality aware. without looking too carefully, i'm pretty sure it will just be the former, that the block that knows the quality context will have access to both sets of mp3s, and can pass the right one in.
4. can you also give me description (or maybe commented out proposed changes to the pairql pairs) of what info you need from the API to wire this all up? remember, the API has lots of good info, if you want it, like for instance it knows the duration in seconds of every mp3.

File: apps/evans/components/core/EmbeddedAudio.tsx [hunk 11]
```diff
@@ -1,60 +1,176 @@
...
+      }
+      setCurrentTime(0);
+      setIsPlaying(true);
+      audio.currentTime = 0;
+    }
+    requestAnimationFrame(whilePlaying);
+  }, [audio, currentTrack, tracks.length]);
+
+  useEffect(() => {
+    setAudio(audioRef.current);
```
[the remark this answers is not in the document]
Comment (kiahjh) [reply apps/evans/components/core/EmbeddedAudio.tsx:43]: It's because `audioRef.current` is initially set to `null`, and putting it in state causes the component to re-render when the jsx hydrates and it gets a real reference to the dom node, so it never gets stuck in a state with a null reference (which was happening prior to this change). There might be a better way to fix that issue though, I'd love to hear if you have another idea

File: apps/evans/components/core/EmbeddedAudio.tsx (hunk 1 above)
Comment (kiahjh) [reply apps/evans/components/core/EmbeddedAudio.tsx:27]: Yeah you're totally right, just removed the callback on every frame and it totally work great. Thanks!

File: apps/evans/components/core/EmbeddedAudio.tsx (hunk 2 above)
Comment (kiahjh) [reply apps/evans/components/core/EmbeddedAudio.tsx:54]: There's no omitted dependency, I just added the `currentTrack.current` so it doesn't get stuck when switching tracks

Comment (kiahjh) [reply apps/evans/components/core/EmbeddedAudio.tsx:54]: eslint deemed it unnecessary; the `exhaustive-deps` thing is a little misleading

File: apps/evans/lib/dates.ts (hunk 8 above)
Comment (kiahjh) [reply apps/evans/lib/dates.ts:44]: I thought it has to do with durations which are sort of times which are sort of dates, but yeah that might be a stretch. Do you want it somewhere else?

File: apps/evans/components/pages/document/ListenBlock.tsx (hunk 6 above)
Comment (kiahjh) [reply apps/evans/components/pages/document/ListenBlock.tsx:12]: Maybe that would be nice...

**Review by jaredh159 — APPROVED** [review]

**Comment by jaredh159** [comment]
thanks. i'm gonna merge, but i disabled vercel deployments temporarily, because i'm going to need to do the API integration work. while i'm at it, i'm going to experiment with the whole ref thing we discussed as well.
